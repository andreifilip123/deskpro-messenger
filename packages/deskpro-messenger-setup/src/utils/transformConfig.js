export default (settings) => {
  const screens = {
    index: {
      screenType: 'Blocks',
      blocks: []
    }
  };

  if (settings.getIn(['chat', 'enabled'])) {
    const chatBlockConfig = settings.getIn(['messenger', 'chat']);
    screens.index.blocks.push({
      blockType: 'StartChatBlock',
      title: chatBlockConfig.get('title', 'Conversation'),
      description: chatBlockConfig.get('description'),
      linkText: chatBlockConfig.get('buttonText', 'Start a chat'),
      showAgentPhotos: chatBlockConfig.get('showAgentPhotos'),
      to: 'startChat'
    });
    screens.startChat = settings.get('chat').toJS();
    screens.startChat.screenType = 'StartChatScreen';
    delete screens.startChat.enabled;
    if (screens.startChat.noAnswerBehavior === 'save_ticket') {
      screens.startChat.ticketFormConfig = [
        {
          fields: [
            {
              name: 'name',
              label: 'Full Name',
              type: 'text',
              validation: ['required'],
              placeholder: 'John Doe'
            },
            {
              name: 'email',
              label: 'Email',
              type: 'text',
              validation: ['required'],
              placeholder: 'john.doe@company.com'
            }
          ]
        }
      ];
    }
  }

  if (settings.getIn(['tickets', 'enabled'])) {
    const ticketBlockConfig = settings.getIn(['messenger', 'tickets']);
    screens.index.blocks.push({
      blockType: 'ScreenLink',
      blockTitle: ticketBlockConfig.get('title'),
      description: ticketBlockConfig.get('description'),
      label: ticketBlockConfig.get('buttonText'),
      to: 'newTicket'
    });
    screens.newTicket = settings.get('tickets').toJS();
    delete screens.newTicket.enabled;
  }

  const enabledGreetings = [null];
  if ('object' === typeof screens.startChat) {
    enabledGreetings.push('/screens/startChat');
  }
  const config = {
    screens,
    themeVars: {
      '--color-primary': settings.getIn(['styles', 'primaryColor'], '#3d88f3'),
      '--color-background': settings.getIn(['styles', 'backgroundColor'], '#3d88f3'),
      '--brand-primary': settings.getIn(['styles', 'primaryColor'], '#3d88f3'),
      '--brand-secondary': settings.getIn(['styles', 'backgroundColor'], '#a9b0b0'),
      '--text-color': settings.getIn(['styles', 'textColor'], '#ffffff'),

    },
    greetings: {},
    enabledGreetings
  };

  return config;
};
