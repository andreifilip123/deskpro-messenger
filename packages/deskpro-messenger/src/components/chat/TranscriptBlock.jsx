import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, FormattedMessage } from 'react-intl';
import * as Yup from 'yup'

import Button from '../form/Button';
import ChatPrompt from '../ui/ChatPrompt';

const schema = Yup.object().shape({
  name:  Yup.string(),
  email: Yup.string().email()
});

const transMessages = {
  questionHeader: {
    id: 'chat.transcript_block.question_header',
    defaultMessage: 'Would you like an email transcript?'
  },
  answerHeader: {
    id: 'chat.transcript_block.answer_header',
    defaultMessage: 'Your transcript is on itâ€™s way to:'
  },
  yesButton: {
    id: 'chat.transcript_block.yes_button',
    defaultMessage: 'Yes'
  },
  noButton: {
    id: 'chat.transcript_block.no_button',
    defaultMessage: 'No'
  },
  sendButton: {
    id: 'chat.transcript_block.send_button',
    defaultMessage: 'Send Message'
  }
};

export class TranscriptBlock extends PureComponent {
  static propTypes = {
    onSend: PropTypes.func.isRequired,
    user: PropTypes.shape({
      name: PropTypes.string,
      email: PropTypes.string
    })
  };

  constructor(props) {
    super(props);
    const { user } = this.props;
    this.state = { viewMode: 'ask', ...user, errors: {} }
  }


  handleYes = (e) => {
    e.preventDefault();
    const { user } = this.props;
    if (user && user.email) {
      this.setState({ viewMode: 'final' }, () => this.submit(user, 1));
    } else {
      this.setState({ viewMode: 'fields' });
    }
  };

  handleNo = (e) => {
    e.preventDefault();
    this.setState({ viewMode: 'hidden' }, () => this.submit({}, 0));
  };

  handleInputChange = (e) => {
    let errors = {};
    const shape = {
      name:  this.state.name,
      email: this.state.email
    };
    shape[e.target.name] = e.target.value;
    try {
      schema.validateSync(shape);
    } catch (e) {
      errors = { [e.path]: e.errors };
    }
    this.setState({ ...shape, errors });
  };

  submit = ({ name, email }, transcript) => {
    this.props.onSend({
      type: 'chat.transcript',
      origin: 'user',
      transcript,
      name,
      email
    });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { name, email, errors} = this.state;
    if (name && email && !errors.email && !errors.name ) {
      this.setState({ viewMode: 'final' }, () => this.submit({name, email}, 1));
    }
  };

  render() {
    const { viewMode } = this.state;
    const { formatMessage } = this.props.intl;

    if (viewMode === 'hidden') {
      return null;
    }

    return (
      <ChatPrompt
        header={formatMessage(
          transMessages[`${viewMode === 'final' ? 'answer' : 'question'}Header`]
        )}
        icon="Notes"
        className={'dpmsg-TranscriptBlock'}
      >
        {viewMode === 'ask' && this.renderYesNo()}
        {viewMode === 'fields' && this.renderFields()}
        {viewMode === 'final' && this.renderFinal()}
      </ChatPrompt>
    );
  }

  renderYesNo() {
    return (
      <div className="dpmsg-PromptContentAgree">
        <button
          className="dpmsg-PromptBtn is-agree"
          name="yes"
          onClick={this.handleYes}
        >
          <FormattedMessage {...transMessages.yesButton} />
        </button>
        <button
          className="dpmsg-PromptBtn is-disagree"
          name="no"
          onClick={this.handleNo}
        >
          <FormattedMessage {...transMessages.noButton} />
        </button>
      </div>
    );
  }

  renderFields() {
    return (
      <div className="dpmsg-GroupInputs">
        <input
          className="dpmsg-Input"
          placeholder="name"
          name="name"
          onChange={this.handleInputChange}
        />
        <input
          className="dpmsg-Input"
          placeholder="email"
          name="email"
          onChange={this.handleInputChange}
        />
        {this.state.errors.email ? <span className="dpmsg-InputError">{ this.state.errors.email.join(', ')}</span> : null}
        <Button width="full" size="medium" onClick={this.handleSubmit}>
          <FormattedMessage {...transMessages.sendButton} />
        </Button>
      </div>
    );
  }

  renderFinal() {
    return (
      <span className="dpmsg-PromptText">
        {this.state.email || this.props.user.email}
      </span>
    );
  }
}

export default injectIntl(TranscriptBlock);
