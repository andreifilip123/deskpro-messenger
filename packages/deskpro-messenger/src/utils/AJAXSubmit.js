// eslint-disable-next-line func-names
const AJAXSubmit = (function () {
  function submitData(oData, config) {
    /* the AJAX request... */
    const oAjaxReq = new XMLHttpRequest();
    oAjaxReq.submittedData = oData;

    if (config.updateProgress) {
      oAjaxReq.addEventListener('progress', config.updateProgress);
    }
    if (config.transferComplete) {
      oAjaxReq.addEventListener('load', config.transferComplete);
    }
    if (config.transferFailed) {
      oAjaxReq.addEventListener('error', config.transferFailed);
    }
    if (config.transferCanceled) {
      oAjaxReq.addEventListener('abort', config.transferCanceled);
    }

    oAjaxReq.onreadystatechange = function() { // Call a function when the state changes.
      if (this.readyState === XMLHttpRequest.DONE) {
        if (this.status >= 400) {
          let message = this.statusText;
          if (this.response && this.response.errors && this.response.errors.file) {
            message = this.response.errors.file;
          }
          config.transferFailed({
            code: this.status,
            message
          });
        }
      }
    }

    oAjaxReq.withCredentials = true;
    /* method is POST */
    oAjaxReq.responseType = 'json';
    oAjaxReq.open('post', oData.receiver, true);
    /* enctype is multipart/form-data */
    const sBoundary = `---------------------------${Date.now().toString(16)}`;
    oAjaxReq.setRequestHeader('Content-Type', `multipart/form-data; boundary=${sBoundary}`);
    if (config.requestHeaders) {
      for (const [name, value] of Object.entries(config.requestHeaders)) {
        oAjaxReq.setRequestHeader(name, value);
      }
    }
    oAjaxReq.send(`--${sBoundary}\r\n${
      oData.segments.join(`--${sBoundary}\r\n`)}--${sBoundary}--\r\n`);
  }

  function processStatus(oData, config = {}) {
    if (oData.status > 0) { return; }
    /* the form is now totally serialized! do something before sending it to the server... */
    // /* doSomething(oData); */
    // /* console.log("AJAXSubmit - The form is now serialized. Submitting..."); */
    submitData(oData, config);
  }

  function pushSegment(oFREvt) {
    this.owner.segments[this.segmentIdx] += `${oFREvt.target.result}\r\n`;
    this.owner.status = this.owner.status - 1;
    processStatus(this.owner, this.config);
  }

  function SubmitRequest(config) {
    this.receiver = config.url;
    this.status = 0;
    this.segments = [];
    for (let nFile = 0; nFile < config.files.length; nFile++) {
      const oFile = config.files[nFile];
      const oSegmReq = new FileReader();
      /* (custom properties:) */
      oSegmReq.segmentIdx = this.segments.length;
      oSegmReq.owner = this;
      oSegmReq.config = config;
      /* (end of custom properties) */
      oSegmReq.onload = pushSegment;
      this.segments.push(`Content-Disposition: form-data; name="file"; filename="${oFile.name}"\r\nContent-Type: ${oFile.type}\r\n\r\n`);
      this.status = this.status + 1;
      oSegmReq.readAsBinaryString(oFile);
    }
    this.segments.push(`Content-Disposition: form-data;`);
    processStatus(this, config);
  }

  // eslint-disable-next-line func-names
  return function (config) {
    if (!config.url) { return; }
    new SubmitRequest(config);
  };
}());

export default AJAXSubmit;
