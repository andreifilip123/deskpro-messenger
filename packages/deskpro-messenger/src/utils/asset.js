window.parent.DESKPRO_MESSENGER_OPTIONS =
  window.parent.DESKPRO_MESSENGER_OPTIONS || {};

export default (path) => {
  let edited = '';
  if (window.parent.DESKPRO_MESSENGER_OPTIONS.baseUrl) {
    edited = window.parent.DESKPRO_MESSENGER_OPTIONS.baseUrl.replace(/\/$/, "");
  }
  return `${edited}/assets/${path}`;
};
