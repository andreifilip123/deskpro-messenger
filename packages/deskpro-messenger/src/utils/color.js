/**
 * Returns true if opposite color is light or not.
 * @param {string} color Color hash.
 * @link https://stackoverflow.com/a/35970186
 */
export const isLightColor = (color) => {
  let hex = color.startsWith('#') ? color.slice(1) : color;
  // convert 3-digit hex to 6-digits.
  if (hex.length === 3) {
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }
  if (hex.length !== 6) {
    throw new Error('Invalid HEX color.');
  }
  const r = parseInt(hex.slice(0, 2), 16),
    g = parseInt(hex.slice(2, 4), 16),
    b = parseInt(hex.slice(4, 6), 16);
  // http://stackoverflow.com/a/3943023/112731
  return r * 0.299 + g * 0.587 + b * 0.114 > 186;
};
