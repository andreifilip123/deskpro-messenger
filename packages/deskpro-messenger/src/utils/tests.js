import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';

const customRender = (node, options) => {
  return render(<IntlProvider locale="en-US">{node}</IntlProvider>, options);
};

// re-export everything
export * from 'react-testing-library';

// override render method
export { customRender as render };
