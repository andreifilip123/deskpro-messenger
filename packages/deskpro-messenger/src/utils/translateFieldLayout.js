/**
 * Translates all strings (labels) in a field layout config.
 *
 * @param {Array} config Field Layout config
 * @param {function} formatMessage Callback from react-intl to translate a string.
 *
 * @returns new translated field layout config.
 */
export default (config, formatMessage) => {
  const t = (str) => formatMessage({ id: str, defaultMessage: str });

  return config.map((layout) => ({
    ...layout,
    fields: layout.fields.map((field) => {
      const newField = { ...field };
      newField.label = t(field.label);
      if (newField.placeholder) {
        newField.placeholder = t(field.placeholder);
      }
      if (Array.isArray(newField.validation)) {
        newField.validation = field.validation.map(
          (rule) =>
            typeof rule === 'object'
              ? { ...rule, message: t(rule.message) }
              : rule
        );
      }
      if (Array.isArray(newField.options)) {
        newField.options = field.options.map((option) => ({
          ...option,
          label: t(option.label)
        }));
      }
      if (
        typeof newField.dataSource === 'object' &&
        Array.isArray(newField.dataSource.getOptions)
      ) {
        newField.dataSource.getOptions = field.dataSource.getOptions.map(
          (option) => ({
            ...option,
            label: t(option.label)
          })
        );
      }
      return newField;
    })
  }));
};
