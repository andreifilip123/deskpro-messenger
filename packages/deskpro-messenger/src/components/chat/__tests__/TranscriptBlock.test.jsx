import React from 'react';
import { render, fireEvent } from '../../../utils/tests';

import TranscriptBlock from '../TranscriptBlock';

describe('<TranscriptBlock />', () => {
  it('should render correctly', () => {
    const onSend = jest.fn();
    const { asFragment } = render(<TranscriptBlock onSend={onSend} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should show form for anon visitor', async () => {
    const onSend = jest.fn();
    const { getByText, getByPlaceholderText, asFragment } = render(
      <TranscriptBlock onSend={onSend} />
    );
    expect(getByText('Yes')).toBeInTheDocument();

    fireEvent.click(getByText('Yes').parentElement);
    expect(getByText('Send Message')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();

    fireEvent.change(getByPlaceholderText('name'), {
      target: { value: 'John Doe' }
    });
    fireEvent.change(getByPlaceholderText('email'), {
      target: { value: 'john.doe@deskpro.com' }
    });
    fireEvent.click(getByText('Send Message').parentElement);
    expect(onSend).toBeCalledWith({
      type: 'chat.transcript',
      origin: 'user',
      name: 'John Doe',
      email: 'john.doe@deskpro.com'
    });
    expect(getByText('Your transcript is on it’s way to:')).toBeInTheDocument();
  });

  it('should send user data without form submittion', () => {
    const onSend = jest.fn();
    const user = {
      name: 'John Doe',
      email: 'john.doe@deskpro.com'
    };
    const { getByText, asFragment } = render(
      <TranscriptBlock onSend={onSend} user={user} />
    );
    expect(getByText('Yes')).toBeInTheDocument();
    fireEvent.click(getByText('Yes').parentElement);
    expect(onSend).toBeCalledWith({
      type: 'chat.transcript',
      origin: 'user',
      name: 'John Doe',
      email: 'john.doe@deskpro.com'
    });
    expect(getByText('Your transcript is on it’s way to:')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not send message for no button', () => {
    const onSend = jest.fn();
    const { getByText, queryByText, container } = render(
      <TranscriptBlock onSend={onSend} />
    );
    expect(getByText('No')).toBeInTheDocument();
    fireEvent.click(getByText('No').parentElement);
    expect(container.firstChild).toBeNull();
    expect(onSend).not.toBeCalled();
    expect(queryByText('Send Message')).toBeNull();
    expect(queryByText('Your transcript is on it’s way to:')).toBeNull();
  });
});
