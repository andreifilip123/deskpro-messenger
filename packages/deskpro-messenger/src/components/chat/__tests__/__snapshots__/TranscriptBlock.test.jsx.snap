// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TranscriptBlock /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="dpmsg-MessagePrompt"
  >
    <div
      class="dpmsg-PromptHeader"
    >
      <i
        class="dpmsg-IconNotes"
      />
      <span
        class="dpmsg-PromptHeaderText"
      >
        Would you like an email transcript?
      </span>
    </div>
    <div
      class="dpmsg-PromptContentAgree"
    >
      <button
        class="dpmsg-PromptBtn is-agdee"
        name="yes"
      >
        <span>
          Yes
        </span>
      </button>
      <button
        class="dpmsg-PromptBtn is-disagree"
        name="no"
      >
        <span>
          No
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TranscriptBlock /> should send user data without form submittion 1`] = `
<DocumentFragment>
  <div
    class="dpmsg-MessagePrompt"
  >
    <div
      class="dpmsg-PromptHeader"
    >
      <i
        class="dpmsg-IconNotes"
      />
      <span
        class="dpmsg-PromptHeaderText"
      >
        Your transcript is on itâ€™s way to:
      </span>
    </div>
    <span
      class="dpmsg-PromptText"
    >
      john.doe@deskpro.com
    </span>
  </div>
</DocumentFragment>
`;

exports[`<TranscriptBlock /> should show form for anon visitor 1`] = `
<DocumentFragment>
  <div
    class="dpmsg-MessagePrompt"
  >
    <div
      class="dpmsg-PromptHeader"
    >
      <i
        class="dpmsg-IconNotes"
      />
      <span
        class="dpmsg-PromptHeaderText"
      >
        Would you like an email transcript?
      </span>
    </div>
    <div
      class="dpmsg-GroupInputs"
    >
      <input
        class="dpmsg-Input"
        name="name"
        placeholder="name"
      />
      <input
        class="dpmsg-Input"
        name="email"
        placeholder="email"
      />
      <button
        class="dpmsg-Button Button-FullWidth Button--medium"
      >
        <span>
          Send Message
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
